include/xtensor/xsort.hpp

```diff
@@ -362,6 +355,402 @@ namespace xt
+            }
+            else
+            {
+                for (std::size_t i = 0; i < n_iters; ++i, ptr += data_secondary_stride, indices_ptr += inds_secondary_stride)
+                {
+                    std::nth_element(indices_ptr, indices_ptr + kth, indices_ptr + last, comp);
+                }
+            }
+        }
+    }
+
+    template <class E, class C, class = std::enable_if_t<!std::is_integral<C>::value, int>>
+    inline auto argpartition(const xexpression<E>& e, const C& kth_container, std::ptrdiff_t axis = -1)
+    {
+        using eval_type = typename detail::sort_eval_type<E>::type;
+        using result_type = typename detail::argsort_result_type<eval_type>::type;
+
+        const auto& de = e.derived_cast();
+
+        std::size_t ax = normalize_axis(de.dimension(), axis);
+
+        if (de.dimension() == 1)
+        {
+            return argpartition<E, C, result_type>(e, kth_container, xnone());
+        }
+
+        C kth_copy = kth_container;
+        if (kth_copy.size() > 1)
+        {
+            std::sort(kth_copy.begin(), kth_copy.end());
+        }
+
+        eval_type ev;
+        result_type res;
+
+        dynamic_shape<std::size_t> permutation, reverse_permutation;
+        bool is_leading_axis = (ax == detail::leading_axis(res));
+
+        if (!is_leading_axis)
+        {
+            std::tie(permutation, reverse_permutation) = detail::get_permutations(de.dimension(), ax, de.layout());
+            ev = transpose(de, permutation);
+        }
+        else
+        {
+            ev = de;
+        }
+        res.resize(ev.shape());
+
+        std::size_t kth = kth_copy.back();
+        detail::argpartition_over_leading_axis(ev, res, kth, -1);
+
+        for (auto it = (kth_copy.rbegin() + 1); it != kth_copy.rend(); ++it)
+        {
+            detail::argpartition_over_leading_axis(ev, res, *it, static_cast<std::ptrdiff_t>(kth));
+            kth = *it;
+        }
+
+        if (!is_leading_axis)
+        {
+            res = transpose(res, reverse_permutation);
+        }
+
+        return res;
+    }
+
+#ifdef X_OLD_CLANG
+    template <class E, class I>
+    inline auto argpartition(const xexpression<E>& e, std::initializer_list<I> kth_container, std::ptrdiff_t axis = -1)
+    {
+        return argpartition(e, xtl::forward_sequence<std::vector<std::size_t>, decltype(kth_container)>(kth_container), axis);
+    }
+#else
+    template <class E, class I, std::size_t N>
+    inline auto argpartition(const xexpression<E>& e, const I(&kth_container)[N], std::ptrdiff_t axis = -1)
+    {
+        return argpartition(e, xtl::forward_sequence<std::array<std::size_t, N>, decltype(kth_container)>(kth_container), axis);
+    }
+#endif
+
+    template <class E>
+    inline auto argpartition(const xexpression<E>& e, std::size_t kth, std::ptrdiff_t axis = -1)
+    {
+        return argpartition(e, std::array<std::size_t, 1>({kth}), axis);
+    }
+
+    template <class E>
+    inline double median(E&& e)
```

Comment:
Should be `floating_promote<typename E::value_type>` (of course `floating_promote` has to be defined).